packages/toolkit/src/query/core/buildMiddleware/polling.ts

```diff
@@ -19,10 +19,24 @@ export const build: SubMiddlewareBuilder = ({
       timeout?: TimeoutId
       pollingInterval: number
     }> = {}
+
     return (next) =>
       (action): any => {
         const result = next(action)
 
+        if (api.internalActions.unsubscribeQueryResult.match(action)) {
+          const { queryCacheKey } = action.payload
+          const existingSubscriptionCount = Object.keys(
+            mwApi.getState()[reducerPath].subscriptions[queryCacheKey] || {}
+          ).length
+
+          // There are no other components subscribed and sharing a poll for this queryCacheKey, so we can
+          // safely remove it
+          if (existingSubscriptionCount === 0) {
+            cleanupPollForKey(queryCacheKey)
+          }
+        }
+
         if (api.internalActions.updateSubscriptionOptions.match(action)) {
```

Comment:
```suggestion
        if (api.internalActions.updateSubscriptionOptions.match(action) || api.internalActions.unsubscribeQueryResult.match(action)) {
```

I'm not sure if I'm missing something, but couldn't this whole PR be replaced by just that single change?